crates/uv-configuration/src/constraints.rs

```diff
@@ -36,16 +38,49 @@ impl Constraints {
     }
 
     /// Apply the constraints to a set of requirements.
+    ///
+    /// NB: Change this method together with [`Overrides::apply`].
     pub fn apply<'a>(
         &'a self,
-        requirements: impl IntoIterator<Item = &'a Requirement>,
-    ) -> impl Iterator<Item = &Requirement> {
+        requirements: impl IntoIterator<Item = Cow<'a, Requirement>>,
+    ) -> impl Iterator<Item = Cow<'a, Requirement>> {
         requirements.into_iter().flat_map(|requirement| {
-            std::iter::once(requirement).chain(
-                self.get(&requirement.name)
-                    .into_iter()
-                    .flat_map(|constraints| constraints.iter()),
-            )
+            let Some(constraints) = self.get(&requirement.name) else {
+                // Case 1: No constraint(s).
+                return Either::Left(std::iter::once(requirement));
+            };
+
+            // ASSUMPTION: There is one `extra = "..."`, and it's either the only marker or part
+            // of the main conjunction.
+            let Some(extra_expression) = requirement
+                .marker
+                .as_ref()
+                .and_then(|marker| marker.top_level_extra())
+                .cloned()
+            else {
+                // Case 2: A non-optional dependency with constraint(s).
+                return Either::Right(Either::Right(
+                    std::iter::once(requirement).chain(constraints.iter().map(Cow::Borrowed)),
+                ));
+            };
+
+            // Case 3: An optional dependency with constraint(s).
+            //
+            // When the original requirement is an optional dependency, the constraint(s) need to
+            // be optional for the same extra, otherwise we activate extras that should be inactive.
```

Comment:
@konstin -- This is really random, but is this necessary for constraints? By adding a constraint, we know that the base package was added as a dependency, right?

Reply:
Also, we already merge markers for constraints.

Reply:
Hmm not sure, i imagine something like:

constraint: `bar <2`

requirement: `root[foo] -> bar`

In that case, we should only add `bar<2` for `root[foo]`, not for `root`.

But it could also be we only need this for overrides.